Details Block: Fix keyboard accessibility issues and allow list view selection to show up inner blocks

yogeshbhutkar opened the pull request:
## What, Why, and How?
Closes #70047

This PR enhances keyboard interaction with the Details block, enabling users to expand inner blocks using the keyboard. It also fixes a bug that previously prevented selecting an inner block from the List View; clicking an inner block in the List View now correctly selects it in the editor canvas.

With this PR, pressing the down arrow key expands the inner blocks within the Details block, while pressing the up arrow key collapses them.

## Testing Instructions

1. Create a post and add the following block markup.

<details>

<summary>Code</summary>

```html
<!-- wp:paragraph -->
<p>First Paragraph Block</p>
<!-- /wp:paragraph -->

<!-- wp:details -->
<details class="wp-block-details">
<summary>Details Block Summary</summary>
<!-- wp:paragraph {"placeholder":"Type / to add a hidden block"} -->
<p>Details Block Content</p>
<!-- /wp:paragraph -->
<!-- wp:paragraph -->
<p>Details Block Content</p>
<!-- /wp:paragraph -->
</details>
<!-- /wp:details -->

<!-- wp:paragraph -->
<p>Second Paragraph Block</p>
<!-- /wp:paragraph -->
```

</details>

2. Try moving through the blocks with the arrow keys.
3. Confirm, the inner blocks within details are accessible using the keyboard.
4. Open the list view and expand the details block.
5. Confirm that clicking on an inner block within the list view opens and selects the inner block properly.

## Screencast

![pr-demo](https://github.com/user-attachments/assets/1b75f22b-5842-41fd-a70a-d1c1a6d93bdc)

yogeshbhutkar committed:
fix: use key down to open/close the details block

yogeshbhutkar committed:
chore: fix import order

yogeshbhutkar commented on packages/block-library/src/details/edit.js:
```diff
@@ -108,7 +128,7 @@ function DetailsEdit( { attributes, setAttributes } ) {
 					/>
 				</summary>
-				{ innerBlocksProps.children }
+				{ shouldRenderChildren && innerBlocksProps.children }
```
When the Details block always displays its children, pressing the up arrow from the block below it doesn't move the selection to the Details block. This approach addresses that issue, but I'm open to a better approach here.

Mamaduka replied:
Depending on the block content, this can be an expensive operation, because it will mount/unmount children. Additionally, it might affect accessibility since the `details` contents are rendered on demand.

yogeshbhutkar replied:
Yes, I had the same concern. I'll try looking at some alternative approaches. For now, I've reverted this.

Mamaduka commented:
Adding handlers for arrow keys introduces a new behavior for `details`, which is expanded using <kbd>Space</kbd> or  <kbd>Enter</kbd> keys. Are we okay with this? It would be good to get a11y feedback on this PR before it's merged. cc @joedolson

P.S. I think @t-hamano was also working on this issue.

Mamaduka commented on packages/block-library/src/details/edit.js:
```diff
@@ -42,6 +44,15 @@ function DetailsEdit( { attributes, setAttributes } ) {
+		[ clientId ]
+	);
+
+	const shouldRenderChildren = hasSelectedInnerBlock || isSelected || isOpen;
```
This reintroduces a buggy behavior resolved via https://github.com/WordPress/gutenberg/pull/67766. We should avoid introducing regression while trying to solve other bugs.

yogeshbhutkar committed:
fix: do not conditionally render details children

alexstine commented:
@yogeshbhutkar Thanks for working on this.

@Mamaduka We need a way to expand the details block because it isn't currently possible in the editor with the keyboard alone. Give it a test on current trunk, not sure how anyone allowed a regression like this to slip by without keyboard testing. Seems like only way is the mouse at this point.

yogeshbhutkar commented:
On a separate note, while spacebar toggling makes sense on the front end, it disrupts the editing experience (I understand it's the default behavior). Disabling this behavior when typing in the summary's rich text would be better. I'd appreciate any accessibility feedback on this.

![bug](https://github.com/user-attachments/assets/0ee34563-6351-4eae-b2d7-83ecc757262b)

Observe, pressing the spacebar while typing toggles the hidden content, which could be confusing for users in the editor.

alexstine commented:
@yogeshbhutkar Yes, absolutely. If you can detect focus in the summary rich text, space keyboard event needs to be prevented for expand/collapse.

t-hamano reviewed (commented):
Thanks for the PR! The approach that uses keyboard events to collapse or expand the content is an interesting approach.

The only concern is that the summary text can span multiple lines. In this case, the content is toggled every time you move from one line to another can be annoying.

Another approach I can think of may be using the Enter key. Additionally, ~let screenreader users know the details block was toggled by the `speak()` method~. In other words, the code would be like this:

```javascript
<summary
	onKeyDown={ ( event ) => {
		if ( event.key === 'Enter' && ! event.shiftKey ) {
			setIsOpen( ! isOpen );
			event.preventDefault();
		}
	} }
>
```

Finally, I think we also need to add some e2e tests to avoid regressions.

t-hamano commented:
Another thing to note is that in CJK languages, an IME (Input Method Editor) is used to input complex characters.

There is [a higher-order function](https://github.com/WordPress/gutenberg/blob/8b7b04ab0c95abd52fb0ef6c7cbb7ec6b440ea23/packages/components/src/utils/with-ignore-ime-events.ts) to ignore keyboard events during IME composition. We will probably need to wrap keydown events in this function in your Details block as well.

See the following PRs for more details:

- https://github.com/WordPress/gutenberg/pull/59081
- https://github.com/WordPress/gutenberg/pull/60090

alexstine commented:
That looks good @t-hamano. Shift enter will insert new line while enter will expand/collapse. I'm not overly thrilled with the UX but it is 100% better than trunk today. Also, can we fix the list view here as well? Any time a hidden block is selected, auto-expand the summary?

Thanks.

t-hamano commented:
@alexstine  

> Also, can we fix the list view here as well? Any time a hidden block is selected, auto-expand the summary?

I think you can confirm that this issue has already been fixed in this PR. This is because when an inner block is selected in the List View, `hasSelectedInnerBlock` is `true` and the `open` attribute of the `details` element is `true`:

https://github.com/WordPress/gutenberg/pull/70056/files#diff-20d83620db8197c679e5617a73e78de6306074dff49588bb00fd5ee35a17793bR105

joedolson commented:
This looks like it's progressing well; thanks for getting on this! I think that you all have this well in hand, but if you need a 2nd review once everything is ready, let me know.

yogeshbhutkar committed:
feat: expose `withIgnoreIMEEvents` to private api

yogeshbhutkar committed:
feat: wrap handler within `withIgnoreIMEEvents`

yogeshbhutkar committed:
chore(test): add e2e test for details block

yogeshbhutkar committed:
fix: prevent space in rich-text from toggling details

yogeshbhutkar committed:
chore: add CHANGELOG

t-hamano commented on packages/block-library/src/details/edit.js:
```diff
@@ -42,6 +48,27 @@ function DetailsEdit( { attributes, setAttributes } ) {
+		}
+	};
+
+	// Prevent spacebar from toggling <details> while typing.
+	const handleSummaryKeyUp = ( event ) => {
+		if ( event.key === ' ' ) {
+			event.preventDefault();
+		}
+	};
```
Do we actually need this code? It seems to work as expected without this code.

yogeshbhutkar replied:
https://github.com/user-attachments/assets/a6dbebaf-faf5-4176-9af7-01787abcacb2

I tested this across different browsers and found that it works correctly only in Firefox. On both Safari and Chrome (macOS), pressing the spacebar while typing in the summary input unexpectedly toggles the hidden blocks. I tried several approaches, but this was the only one that reliably worked. Additionally, when I run the Playwright test for this case locally, it consistently fails if this code is removed.

t-hamano replied:
Interesting. In my environment ( Chrome and Firefox on Windows 11), the space key doesn't toggle the hidden blocks in the editor. The behavior may vary depending on the OS.

We also need to wrap this function with the `withIgnoreIMEEvents`, as in an IME, we use the space key during an IME composition, not just `enter` and arrow keys:

https://github.com/user-attachments/assets/102a89b3-226a-4a84-ac10-564749e32260

yogeshbhutkar replied:
I've wrapped the function with `withIgnoreIMEEvents` in the latest commit. The docs for `withIgnoreIMEEvents` only mention key down events and not up events. (It needs to be generalized.)

Ref.
https://github.com/WordPress/gutenberg/blob/8b7b04ab0c95abd52fb0ef6c7cbb7ec6b440ea23/packages/components/src/utils/with-ignore-ime-events.ts#L9

MDN Reference:
https://developer.mozilla.org/en-US/docs/Web/API/Element/keyup_event#keyup_events_with_ime

Should the docs be updated?

t-hamano replied:
According to the test results by @megane9988, it seems to work as expected without wrapping the function with `withIgnoreIMEEvents`. We can probably remove `withIgnoreIMEEvents` from the `onKeydown` event handler as well. Sorry for the confusion.

> The docs for `withIgnoreIMEEvents` only mention key down events and not up events. (It needs to be generalized.)

Good catch, we can address this issue as a separate PR.

t-hamano commented on packages/components/src/private-apis.ts:
```diff
@@ -18,5 +19,6 @@ lock( privateApis, {
 	Theme,
 	Menu,
 	kebabCase,
+	withIgnoreIMEEvents,
```
@WordPress/gutenberg-components 

We export this HOF as a private API. We need this function in the block library to prevent the Details content from being toggled by the Enter key event fired by the IME event.

t-hamano replied:
Update: Let's test whether it works without this HOF.

t-hamano replied:
Update: If the HOF is not used, unintended events occur in the IME and it does not work properly (See https://github.com/WordPress/gutenberg/pull/70056#issuecomment-2862592442). Therefore, it is necessary to use HOF as a private API.

t-hamano commented on test/e2e/specs/editor/blocks/details.spec.js:
```diff
@@ -0,0 +1,84 @@
+			attributes: {
+				summary: 'Details summary',
+			},
+		} );
```
```suggestion
		await editor.insertBlock( {
			name: 'core/details',
			attributes: {
				summary: 'Details summary',
			},
			innerBlocks: [
				{
					name: 'core/paragraph',
					attributes: { content: 'Details content' },
				},
			],
		} );
```

It might be better to use an inner block with explicit text.

t-hamano commented on test/e2e/specs/editor/blocks/details.spec.js:
```diff
@@ -0,0 +1,84 @@
+		} );
+	} );
+
+	test( 'typing space in summary rich-text should not toggle details', async ( {
+		editor,
+	} ) => {
```
I'm not sure if this test is necessary. See https://github.com/WordPress/gutenberg/pull/70056#discussion_r2078800049

yogeshbhutkar replied:
As it's established that this behavior can be [OS specific](https://github.com/WordPress/gutenberg/pull/70056#discussion_r2078852443), I believe it's fair to keep this e2e test.

t-hamano reviewed (commented):
Thanks for the update, @yogeshbhutkar!

I think this PR is going pretty well.

> Any time a hidden block is selected, auto-expand the summary?

Can we add an e2e test for this? The steps should be like this:

- Insert a Details block
- Open the List view
- Press the right arrow key: The inner blocks of the Details block are displayed in the List View
- Press the down arrow key: The first inner block is focused in the List View
- Press the enter key: The first inner block should be focused in the editor canvas

yogeshbhutkar committed:
fix: wrap key up handler with `withIgnoreIMEEvents`

megane9988 commented:
I tested it.

OS: macOS 15.4.1（24E263）
Browser: Chrome 135.0.7049.115
IME: Google 日本語入力 ( https://www.google.co.jp/ime/ )

# Test Steps
- Adding a Details Block
-  Pressing the Enter key on the Summary text should toggle the content open and closed.
-  Pressing Enter + Shift on the Summary text should insert a line break without toggling the content.
-  Key operations while inputting Japanese (such as the Enter key and Space key) should be ignored, and the content should not toggle.


https://github.com/user-attachments/assets/087c597e-7755-4e68-9ed5-61c7cb419ce4

It worked well.

megane9988 commented:
I tested it.

OS: macOS 15.4.1（24E263）
Browser: Safari 18.4 (20621.1.15.11.10)
IME: Google 日本語入力 ( https://www.google.co.jp/ime/ )

# Test Steps
- Adding a Details Block
-  Pressing the Enter key on the Summary text should toggle the content open and closed.
-  Pressing Enter + Shift on the Summary text should insert a line break without toggling the content.
-  Key operations while inputting Japanese (such as the Enter key and Space key) should be ignored, and the content should not toggle.


https://github.com/user-attachments/assets/5388c2d0-df73-4a87-9335-fb2c88c22995


It also worked well.

megane9988 commented:
I also tested using ことえり ( macOS default IME ) in both browsers (Chrome and Safari).
It also worked well.

https://github.com/user-attachments/assets/aa5d0822-f330-466e-9b58-1a333a4992ee

yogeshbhutkar committed:
chore(e2e): updated tests

* Adds using inner block content wherever required
* Adds an e2e test to verify list view auto expand and focus

yogeshbhutkar committed:
chore: move `CHANGELOG` to unreleased

yogeshbhutkar commented:
I've incorporated all the feedback. A few follow-ups for the details block (separate from this PR):

1. We should add e2e coverage for drag-and-drop, as it previously caused a [regression](https://github.com/WordPress/gutenberg/pull/67766) and remains a risk.

2. Keyboard navigation should also be possible when the block is collapsed. As of now, hidden blocks restrict the movement, and removing them might not be good for accessibility.

3. When inserting inner blocks, only the first shows the correct `/` placeholder (Type `/` to add a hidden block). Subsequent ones revert to the default. We should persist the template message while adding hidden blocks to avoid confusion. Coming out of details requires double enter, but the editor screen shows no feedback. Plus, if we are adding a hidden block, the message should always state so.

I'd like to know the opinion of others on this.

yogeshbhutkar committed:
chore: revert usage of `withIgnoreIMEEvents`

t-hamano commented:
@yogeshbhutkar Thanks for the update! 

I think it's almost ready to ship. The only thing to check is whether it works properly with CJK languages ​​without using `withIgnoreIMEEvents`.

@megane9988 Could you please test this PR from the following link one more time to make sure it works properly? 🙏

https://playground.wordpress.net/?gutenberg-pr=70056

@joedolson @alexstine I think this PR is ready from a code perspective, let me know if there are any other accessibility issues that should be addressed.

megane9988 commented:
@t-hamano 
Unfortunately, it did not work well.

I tested it.
https://playground.wordpress.net/?gutenberg-pr=70056

OS: macOS 15.4.1（24E263）
Browser: Safari 18.4 (20621.1.15.11.10) and Chrome 135.0.7049.115
IME: ことえり ( macOS default IME ) 

# Test Steps

- Pressing the Enter key on the Summary text should toggle the content open and closed.
- Pressing Enter + Shift on the Summary text should insert a line break without toggling the content.
- Key operations while inputting Japanese (such as the Enter key and Space key) should be ignored, and the content should not toggle.

# Result
- ✅️ Pressing the Enter key on the Summary text should toggle the content open and closed.
- ✅️ Pressing Enter + Shift on the Summary text should insert a line break without toggling the content.
- ❌️ Key operations while inputting Japanese (such as the Enter key and Space key) should be ignored, and the content should not toggle.
    - A Space key worked well. But the Enter key did not work well. It also toggles open and close when I press the Enter key.

https://github.com/user-attachments/assets/1fc4f176-2f8b-4200-9194-2b0fc126ebf2

yogeshbhutkar commented:
Based on the test results, I think it's best to at least wrap the keydown handler function. The space would always work because it technically just prevents the default behavior; therefore, no abnormalities can be seen in the editor.

What do you think, @t-hamano?

t-hamano commented:
Thank you so much for testing, @megane9988!

> I think it's best to at least wrap the keydown handler function.

@yogeshbhutkar Yes, based on his tests, it would be enough to apply HOF only to the keydown handler. Sorry for the frequent suggestions for changes.

yogeshbhutkar committed:
fix: apply `withIgnoreIMEEvents` to key down

Manual testing with IMEs showed that the HOF can safely be applied just to the key down events

t-hamano reviewed (approved):
LGTM! Before shipping this PR, let's get some accessibility feedback again.

alexstine commented:
@yogeshbhutkar This is looking much better. One final thought. Since it isn't clear that the details can be expanded, maybe add `aria-describedby` with some basic instructions?

> Press Enter to toggle details child blocks.

The expanded/collapsed state does enough to notify users of the exposed content, nice job on that.

I also agree it would be nice to do something about the "hidden" in the new block placeholder phrasing. The block isn't technically hidden if the details element is expanded.

Thanks.

yogeshbhutkar committed:
feat: add visually hidden instructions for expanding/collapsing details

yogeshbhutkar commented:
Thanks for the feedback, @alexstine.

I've implemented the suggested feedback. When you have a moment, could you please re-review the changes and approve them if everything looks good?

alexstine commented:
@yogeshbhutkar Can you add the `aria-describedby` to the input rich text field, not the block wrapper? If you press Enter from the block wrapper, you will insert a new block, not expand the element.

It also seems like something has changed where the Up/Down arrow keys can now toggle the display and this can be rather disruptive as Up/Down arrow keys allow screen reader users to read by line.

Thanks.

alexstine commented on packages/block-library/src/details/edit.js:
```diff
@@ -30,10 +38,13 @@ const TEMPLATE = [
-	const blockProps = useBlockProps();
+	const instanceId = useInstanceId( DetailsEdit, 'details-edit-desc' );
+	const blockProps = useBlockProps( {
+		'aria-describedby': instanceId,
```
This cannot be defined here as pressing Enter at block wrapper level will insert a new block, not expand/collapse the summary.

yogeshbhutkar replied:
Thanks! I've moved this to `RichText` and verified that the message now triggers correctly when typing in the `RichText` field.

alexstine reviewed (commented):
I'm not sure if `RichText` can take `aria-describedby` by default, hopefully so. I would also like to see if anyone else can reproduce the Up/Down arrow keys toggling the state of the hidden blocks while in the summary edit field.

yogeshbhutkar committed:
fix: add `aria-describedby` and add it to RichText

yogeshbhutkar commented:
> I would also like to see if anyone else can reproduce the Up/Down arrow keys toggling the state of the hidden blocks while in the summary edit field.

I tested this and wasn't able to toggle the hidden blocks using the arrow keys; only the `Enter` key worked, as expected. I've attached a screencast for reference.

cc: @alexstine

https://github.com/user-attachments/assets/a2b2eb41-c123-4b38-955f-18242ea3e2bd

t-hamano commented:
Adding the `VisuallyHidden` component inside the `summary` element seems to add unexpected focus.

For example, add a Details block to the top and focus on the post title. The screen reader will read it as follows:

1. Press the down key: `Block: Details  document  blank`
2. Press the down key: `Press Enter to expand or collapse the details. Summary Text  button  collapsed  1 of 1`
3. Press the down key: `Write summary  edit  multi line  Press Enter to expand or collapse the details.  Summary Text`

https://github.com/user-attachments/assets/36bc4758-16d2-455d-ab24-24cebfb3c794

A second focus should not be necessary.

I would suggest that since the `RichText` component already has an `aria-label`, we just add the text there.

```diff
diff --git a/packages/block-library/src/details/edit.js b/packages/block-library/src/details/edit.js
index a21a00bade..c2df5721e9 100644
--- a/packages/block-library/src/details/edit.js
+++ b/packages/block-library/src/details/edit.js
@@ -130,14 +130,11 @@ function DetailsEdit( { attributes, setAttributes, clientId } ) {
                                        onKeyDown={ withIgnoreIMEEvents( handleSummaryKeyDown ) }
                                        onKeyUp={ handleSummaryKeyUp }
                                >
-                                       <VisuallyHidden id={ instanceId }>
-                                               { __(
-                                                       'Press Enter to expand or collapse the details.'
-                                               ) }
-                                       </VisuallyHidden>
                                        <RichText
                                                identifier="summary"
-                                               aria-label={ __( 'Write summary' ) }
+                                               aria-label={ __(
+                                                       'Write summary. Press Enter to expand or collapse the details.'
+                                               ) }
                                                aria-describedby={ instanceId }
                                                placeholder={ placeholder || __( 'Write summary…' ) }
                                                withoutInteractiveFormatting
```

So the screen reader should read something like this:

1. Press the down key: `Block: Details  document  blank`
2. Press the down key: `Write summary. Press Enter to expand or collapse the details.  edit  multi line  Summary Text`

> I would also like to see if anyone else can reproduce the Up/Down arrow keys toggling the state of the hidden blocks while in the summary edit field.

In the first state of this PR, we were using the up and down arrow keys.
But now we're using the enter key to toggle the content, so I think this issue is no longer a problem.

yogeshbhutkar committed:
fix: update aria-label for RichText

yogeshbhutkar committed:
fix: remove the usage of `aria-describedby`

t-hamano commented:
I think this PR is ready to ship. @alexstine What do you think?

I believe the following requirements are met in this PR:

- When the `RichText` component gets focus, the screen reader reads out the appropriate description.
- Pressing the Enter key toggles content, but is ignored during IME composition.
- Pressing the Shift+Enter key creates a multiline.
- When a block inside a Details block is selected in the List View, the content expands and the block gets focused.

yogeshbhutkar committed:
chore: move `CHANGELOG` to unreleased

yogeshbhutkar commented:
This PR appears to have gone stale. Given that we've already completed the initial round of accessibility feedback and considering that this PR clearly enhances accessibility overall, I'd suggest proceeding with this PR. Any remaining or future accessibility concerns can be addressed as a follow-up.

cc: @t-hamano

t-hamano commented:
> This PR appears to have gone stale. Given that we've already completed the initial round of accessibility feedback and considering that this PR clearly enhances accessibility overall, I'd suggest proceeding with this PR. Any remaining or future accessibility concerns can be addressed as a follow-up.

I agree. This PR is much better than the trunk one, so if there are improvements to the label reading, let's address them in a follow-up.

joedolson commented:
Thanks for moving forward with this! I definitely agree that this is much better than the trunk version, and it makes far more sense to ship the improvements and iterate, rather than continue to need to refactor the patch when it's gone out of date.

t-hamano commented:
I am checking the PRs with the "Backport to WP Minor Release" label applied to clarify which PRs should be backported to [the WordPress 6.8.2 release](https://make.wordpress.org/core/2025/06/17/wordpress-6-8-2-release-schedule/).

I believe this PR should be backported to the 6.8.2 release since it is a regression that occurred in 6.8.

joedolson commented:
Agreed!

t-hamano commented:
This PR was cherry-picked into the `wp/6.8` branch for the WP 6.8.2 release: #70557
